carla/recourse_methods/catalog/roar/model.py

```diff
@@ -0,0 +1,273 @@
+                raise ValueError(
+                    "Incorrect shape of coefficients. Expected shape: (num_features,)"
+                )
+            if not (len(intercepts.shape) == 0):
+                raise ValueError("Incorrect shape of coefficients. Expected shape: ()")
+
+            # Reshape to desired shape: (num_of_instances, num_of_features)
+            coeffs = np.vstack([self._coeffs] * factuals.shape[0])
+            intercepts = np.vstack([self._intercepts] * factuals.shape[0]).squeeze(
+                axis=1
+            )
+        else:
+            if self._mlmodel.model_type == "linear":
+                coeffs_neg = self._mlmodel.raw_model.output.weight.detach()[0].numpy()
+                coeffs_pos = self._mlmodel.raw_model.output.weight.detach()[1].numpy()
+
+                intercepts_neg = np.array(
+                    self._mlmodel.raw_model.output.bias.detach()[0].numpy()
+                )
+                intercepts_pos = np.array(
+                    self._mlmodel.raw_model.output.bias.detach()[1].numpy()
+                )
+
+                self._coeffs = -(coeffs_neg - coeffs_pos)
+                self._intercepts = -(intercepts_neg - intercepts_pos)
+
+                # Local explanations via LIME generate coeffs and intercepts per instance, while global explanations
+                # via input parameter need to be set into correct shape [num_of_instances, num_of_features]
+                coeffs = np.vstack([self._coeffs] * factuals.shape[0])
+                intercepts = np.vstack([self._intercepts] * factuals.shape[0]).squeeze(
+                    axis=1
+                )
+            elif self._mlmodel.model_type == "ann":
+                log.info("Start generating LIME coefficients")
+                coeffs, intercepts = self._get_lime_coefficients(factuals)
+                log.info("Finished generating LIME coefficients")
+            else:
+                raise ValueError(
+                    f"Model type {self._mlmodel.model_type} not supported in ROAR recourse method"
+                )
+
+        cfs = []
+        for index, row in factuals.iterrows():
+
+            coeff = coeffs[index]
+            intercept = intercepts[index]
+
+            counterfactual = roar_recourse(
+                self._mlmodel.raw_model,
+                row.to_numpy().reshape((1, -1)),
+                coeff,
+                intercept,
+                cat_features_indices,
+                binary_cat_features=self._binary_cat_features,
+                feature_costs=self._feature_costs,
+                lr=self._lr,
+                lambda_param=self._lambda_param,
+                delta_max=self._delta_max,
+                y_target=self._y_target,
+                norm=self._norm,
+                t_max_min=self._t_max_min,
+                loss_type=self._loss_type,
+                loss_threshold=self._loss_threshold,
```

Comment:
Rather then passing all those attributes, it might be cleaner to just pass the hyper-parameters dictionary.